File: Niveau 3/Niveau 3.11/amis.py
#-------------------------------------------------------------------------------
# Name:        Amis d’amis
# Purpose:     France-IOI niveau 3.11
#
# Author:      Sébastien
#
# Created:     30/10/2022
#-------------------------------------------------------------------------------

import sys
input = sys.stdin.readline

def main():
    i = int(input())
    n = int(input())
    relations = []
    distance = [-1]*65000 # degré d'amitié : 0 = soi-même, 1 = ami direct, 2 = ami d'ami, -1 = pas ami
    distance[i] = 0
    for _ in range(n):
        i1, i2 = map(int, input().split())
        if i1 == i:
            distance[i2] = 1
        elif i2 == i:
            distance[i1] = 1
        else:
            relations.append((i1, i2))
            relations.append((i2, i1))

    nbAmis = 0
    for i1, i2 in relations:
        if  distance[i1] == 1 and distance[i2] == -1:
##            print("ami", i2)
            nbAmis += 1
            distance[i2] = 2

    print(nbAmis)

if __name__ == '__main__':
    main()

"""
0
11
0 1
0 2
0 3
1 2
1 3
2 3
4 1
2 5
2 4
3 6
7 8
"""

File: Niveau 3/Niveau 3.11/amis.hs
--------------------------------------------------------------------------------
-- Name:        Amis d’amis
-- Purpose:     France-IOI niveau 3.11
--
-- Author:      Sébastien
--
-- Created:     31/10/2022
--------------------------------------------------------------------------------

import Data.Set (insert, empty, member, notMember)

main :: IO ()
main = do
    i <- readLn :: IO Int
    _ <- readLn :: IO Int
    lignes <- fmap lines getContents
    let relations = foldr (\l r -> let (i1:i2:_) = map read . words $ l :: [Int]
                                   in (i1,i2):(i2,i1):r) [] lignes
        amis = foldr (\(i1, i2) s -> if i1 == i then insert i2 s else s) empty relations
        amis2 = foldr (\(i1, i2) s -> if i2 /= i && i1 `member` amis && i2 `notMember` amis then insert i2 s else s) empty relations
    print $ length amis2
    
    